org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,63 +16,106 @@ namespace MixedReality.Toolkit.Input
+        /// interactable stopped being detected and update their buttons front plate RawImage.
+        /// </summary>
+        private readonly HashSet<IXRProximityInteractable> previouslyDetectedInteractables = new();
 
         /// <summary>
-        /// Used to keep track of the previously detected colliders so that we can know which
-        /// colliders stopped being detected and update their buttons front plate and rounded
-        /// rect if they are labeled as dynamic (based on proximity).
+        /// Stores the currently detected interactables (updated every frame).
         /// </summary>
-        List<Collider> previouslyDetectedColliders = new List<Collider>();
+        private readonly List<IXRProximityInteractable> currentlyDetectedInteractables = new();
+
+        /// <summary>
+        /// Stores the interactables that are no longer detected (updated every frame).
+        /// </summary>
+        private readonly List<IXRProximityInteractable> noLongerDetectedInteractables = new();
 
         /// <inheritdoc />
         public override bool IsModeDetected()
         {
             bool result = base.IsModeDetected() || IsNearInteractorSelecting();
             if (result)
             {
-                for (int i = 0; i < previouslyDetectedColliders.Count; i++)
+                UpdateCurrentlyDetectedInteractables();
+                UpdateProximityExited();
+                UpdateProximityEntered();
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Calls OnProximityExited for all interactables that were previously ProximityHover detected but are no longer detected.
+        /// </summary>
+        private void UpdateProximityExited()
+        {
+            noLongerDetectedInteractables.Clear();
+
+            foreach (IXRProximityInteractable previouslyDetectedInteractable in previouslyDetectedInteractables)
+            {
+                if (!currentlyDetectedInteractables.Contains(previouslyDetectedInteractable))
+                {
+                    noLongerDetectedInteractables.Add(previouslyDetectedInteractable);
+                }
+            }
+
+            foreach (IXRProximityInteractable noLongerDetectedInteractable in noLongerDetectedInteractables)
+            {
+                noLongerDetectedInteractable.OnProximityExited(new ProximityHoverExitedEventArgs(noLongerDetectedInteractable));
+                currentlyDetectedInteractables.Remove(noLongerDetectedInteractable);
+            }
+        }
+
+        /// <summary>
+        /// Call OnProximityEntered for all interactables that are currently detected but were not detected previously.
+        /// </summary>
+        private void UpdateProximityEntered()
+        {
+            foreach (IXRProximityInteractable currentlyDetectedInteracable in currentlyDetectedInteractables)
+            {
+                if (previouslyDetectedInteractables.Add(currentlyDetectedInteracable))
```

Comment:
Do we ever need to clear or remove from `previouslyDetectedInteractables`? On a quick readthrough, it feels like we'll only send `OnProximityEntered` the first time an interactable is hovered because it'll stay in this list forever.

Reply:
Oh, is the line above where I asked about `currentlyDetectedInteractables.Remove` supposed to be `previouslyDetectedInteractables.Remove`?

Reply:
Yes, you are correct, I missed that.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/72b8f65ef00615ff47b49f537b66da5c33bc1e23, resolving this.